text/2020-02-04-auth.md

```diff
@@ -0,0 +1,228 @@
+# Auth for applications
+
+## Summary
+
+TiKV authentication and authorization can be implemented by leveraging
+
+* key space registration
+* token signing
+* delegation
+
+PD maintains key spaces and signs tokens for clients that authenticate.
+TiKV verifies the tokens.
+Clients may delegate their tokens to permit direct TiKV access for users in their own auth system.
+
+This can solve the most pressing auth concerns without attempting to
+
+* push down TiDB's auth into TiKV
+* develop a separate RBAC in TIKV
+
+
+## Motivation
+
+TiKV currently has no concept of authentication.
+TiDB supports authentication.
+This has been done in the TiDB layer probably in partly because it is a large effort to achieve MySQL compatible auth mechanisms.
+However, TiKV exposes its data over a network connection, whereas MySQL storage does not.
+
+This means that any attacker that can connect to a TiKV node can exfiltrate all its data.
+This risk can be mitigated by the usage of TiKV's support for mutual TLS authentication.
+TLS can identify a client by its certificate, but without auth in TiKV, this just verifies that it came from a trusted CA.
+So every connected client is effectively a root user.
+There is also risk that the certificate issuance is not properly closed to TiKV clients since we don't identify the clients that connect.
+
+We want TiKV to be used as a stand-alone project.
+This means that TiKV network access to more diverse clients than just PD/TiDB is a requirement.
+Essentially every client must implement their own auth just as TiDB has done.
+And all these client machines with the ability to access TiKV must be trusted as root users.
+
+
+## Detailed design
+
+#### Key Spaces and user spaces
+
+I first want to establish the concept of a “key space”.
+TiDB is one potential user of TiKV.
+One also might want to run 2 entirely separate TiDB clusters against the same TiKV cluster or run other applications against TiKV.
```

Comment:
can we avoid 1 TiDB use 2 "key space"?

Reply:
I am not sure what this question is asking, perhaps you can add more detail? I would expect 1 TiDB to use 1 key space. This auth proposal would ensure it accesses only its key space.

Reply:
Yes, I mean it will be more complex if we have 1 TiDB use multiple TiKV key space(TiDB maybe need choose different user to access different key ranges).

if we expect 1 TiDB to use 1 key space that will be ok :D

Reply:
If TiDB had multiple key spaces it would use a different key space for each MySQL schema/database or perhaps for every table. But I don't think there is actually any benefit to this approach if the delegation with restrictions that is proposed here is implemented.

I changed the proposal now to only briefly mention the key space proposal.